service/storage_service.cc

```diff
@@ -5880,6 +5886,87 @@ future<> storage_service::add_tablet_replica(table_id table, dht::token token, l
     });
 }
 
+future<> storage_service::del_tablet_replica(table_id table, dht::token token, locator::tablet_replica dst, loosen_constraints force) {
+    auto holder = _async_gate.hold();
+
+    if (this_shard_id() != 0) {
+        // group0 is only set on shard 0.
+        co_return co_await container().invoke_on(0, [&] (auto& ss) {
+            return ss.del_tablet_replica(table, token, dst, force);
+        });
+    }
+
+    while (true) {
+        auto guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
+
+        while (_topology_state_machine._topology.is_busy()) {
+            const auto tstate = *_topology_state_machine._topology.tstate;
+            if (tstate == topology::transition_state::tablet_draining ||
+                tstate == topology::transition_state::tablet_migration) {
+                break;
+            }
+            rtlogger.debug("del_tablet_replica(): topology state machine is busy: {}", tstate);
+            release_guard(std::move(guard));
+            co_await _topology_state_machine.event.wait();
+            guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
+        }
+
+        std::vector<canonical_mutation> updates;
+        auto& tmap = get_token_metadata().tablets().get_tablet_map(table);
+        auto tid = tmap.get_tablet_id(token);
+        auto& tinfo = tmap.get_tablet_info(tid);
+        auto last_token = tmap.get_last_token(tid);
+        auto gid = locator::global_tablet_id{table, tid};
+
+        if (tmap.get_tablet_transition_info(tid)) {
+            throw std::runtime_error(fmt::format("Tablet {} is in transition", gid));
+        }
+        auto* node = get_token_metadata().get_topology().find_node(dst.host);
+        if (!node) {
+            throw std::runtime_error(format("Unknown host: {}", dst.host));
+        }
+        if (dst.shard >= node->get_shard_count()) {
+            throw std::runtime_error(format("Host {} does not have shard {}", *node, dst.shard));
+        }
+
+        if (!locator::contains(tinfo.replicas, dst.host)) {
+            throw std::runtime_error(fmt::format("Tablet {} doesn't have replica on {}", gid, dst.host));
+        }
+
+        locator::tablet_replica_set new_replicas;
+        new_replicas.reserve(tinfo.replicas.size() - 1);
+        std::copy_if(tinfo.replicas.begin(), tinfo.replicas.end(), std::back_inserter(new_replicas), [&dst] (auto r) { return r != dst; });
+
+        updates.emplace_back(replica::tablet_mutation_builder(guard.write_timestamp(), table)
+            .set_new_replicas(last_token, new_replicas)
+            .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
+            .set_transition(last_token, locator::tablet_transition_kind::rebuild)
```

Comment:
We are removing a replica for a given tablet. But why do we in the transition of rebuild?

Reply:
Because there are only two transition kinds -- migrate and rebuild. Adding a replica was done with the help of rebuild, removing a replica just continues this "trend". It's possible to del replica with "migrate" kind as well, but I kept "rebuild" for symmetry with add_replica. Maybe it makes sense to add 3rd kind, but so far the rebuild kind just works

Reply:
I guess the logic is the process is the same and can be reused. The meaning of "rebuild" transition is "change the replica set, rebuilding new replicas from existing ones". If there are no new replicas, nothing is rebuilt.

Reply:
Wouldn't that trigger the on_internal_error() in get_migration_streaming_info() due to missing pending replica?

Reply:
It would, but the rest of the code ~~is~~ should've been patched by this PR not to call get_migration_streaming_info()

Reply:
Making it symmetry with add_replica makes sense to me. However, the word "rebuild" is overloaded too much. It is very confusing with nodetool rebuild ops already. People also use the word rebuild when they are starting a new node to replace a dead node. 

Perhaps:

tablet_transition_kind:migration   : moves tablet around for load balance purpose
tablet_transition_kind::sync           :  add remove replicas              (sync is not a good word either, open to better ideas)
tablet_transition_kind:repair         :  fix the mismatch between replicas

However, we have a tablet_draining track outside the tablet_migration track, which is not very symmetric. 
@tgrabiec can we move the tablet_draining to tablet_migration track somehow. If we do, all the tablet work will be in the same track.

Reply:
> But the load balancer can theoretically call it at any time for every transition to compute the current load.

Right, missed that. Will fix

Reply:
@tgrabiec  https://github.com/scylladb/scylladb/pull/18203/files#r1565033102 Why do we put tablet_draining track outside the tablet_migration track? Can we merge them into a single topology::transition_state state?

Reply:
The behavior of the drain track is different. For drain we don't want it to be preemptable, and we want to jump to a different stage after it's done, not release the state machine.

Reply:
What do you mean by preemptable exactly? Who preempt who in what condition?

Can't we jump to whatever tablet_drain jumps to after the migration tablet_migration is done for a special tablet_migration kind?

Reply:
> What do you mean by preemptable exactly? Who preempt who in what condition?

Preemption means that tablet migration track will end sooner when there is a preempting event - a topology request is pending. We preempt by stopping scheduling new migrations, already started migrations will still need to complete. Drain is not preemptable - topology request is already active during drain and we want to end drain when all affected tablets are migrated away from the drained node.  

> 
> Can't we jump to whatever tablet_drain jumps to after the migration tablet_migration is done for a special tablet_migration kind?

Drain is not using special transition kind. It schedules regular migrations. They are not scheduled before entering drain. It can be that there are no affected tablets which need to be drained. 

Maybe you could recover this information from the state of topology, but this would be error prone. What we could do is to put the information that this is a drain in another field in topology state machine, which would be picked up by tablet migration track.